[Distributed] add FP8 support to NaN checker

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135961
* __->__ #135891

Adding support for `torch.float8_e4m3fn` and `torch.float8_e5m2`

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[Distributed] add FP8 support to NaN checker

[ghstack-poisoned]

rghosh08 commented:
Preparing review...

rghosh08 commented:
## PR Reviewer Guide 🔍

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>PR contains tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Key issues to review</strong><br><br>

<a href='https://github.com/pytorch/pytorch/pull/135891/files#diff-9f639571a250cffbe9cded7d2fbb5ad6311e4be9c0c7610e5ba85930806e7f38R370-R373'><strong>Code Smell</strong></a><br>The comment in line 370 suggests that filling values into a FP8 tensor is currently not supported. This might be worth considering for future improvements.

<a href='https://github.com/pytorch/pytorch/pull/135891/files#diff-4252ff3f4c2f04512460c2f5648ffd8b3deac993a5e0183eac2431b1da010454R168-R173'><strong>Code Smell</strong></a><br>The AT_DISPATCH_FLOATING_TYPES_AND4 macro now includes support for FP8 types. However, the comment above it still mentions support for only Half and BFloat16 types. This might be worth updating for consistency.

</td></tr>
</table>

wconstab reviewed (approved):
LGTM, except for missing optimized template for the f8 kernel.  Can you determine if its worth adding before deciding to land this as is or optimize further?

rghosh08 commented:
> deciding

sure. I can look into this.

kwen2501 committed:
Update on "[Distributed] add FP8 support to NaN checker"


Adding support for `torch.float8_e4m3fn` and `torch.float8_e5m2`

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
